Fix CRD status subresource ratcheting

JoelSpeed opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

This PR updates the CRD status subresource storage to implement the same validation ratcheting as is implemented in the main CRD storage implementation.

It updates the ratcheting validation integration tests to apply each of the test cases both to the main CR and the CR's status subresource, to prove the validation is consistent across the two.

There are two exceptions to this. 

One of the CEL tests does not make sense for status (it expects a `null` oldSelf which is not possible for a status subresource) so I've added a partial skip for that specific test, so that it doesn't attempt to run against the status subresource.

The second adds a transition rule, that "should not be ratcheted", and fails when updating other fields. Since the tests would normally add the same validations to both spec and status in the way I've done the testing here, this creates a deadlock, since the spec write errors on the status rule, and the status write errors on the spec rule. So this on is also skipped for now.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #129503

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
Enables ratcheting validation on status subresources for CustomResourceDefinitions
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

JoelSpeed committed:
Fix status subresource ratcheting

JoelSpeed commented:
/test pull-kubernetes-integration

JoelSpeed commented:
/test pull-kubernetes-unit
/test pull-kubernetes-verify
/test pull-kubernetes-e2e-kind

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go:
```diff
@@ -95,6 +97,12 @@ type expectError struct {
+		// and also when updating the CR status subresource.
+		if agg, ok := err.(utilerrors.Aggregate); ok && len(agg.Errors()) != 2 {
+			return fmt.Errorf("expected 2 errors, got %v", len(agg.Errors()))
+		}
```
this extra logic has some smell. `expectError` does not sound like it will check that there are two errors aggregated.

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go:
```diff
@@ -228,10 +279,21 @@ func (u updateMyCRDV1Beta1Schema) Do(ctx *ratchetingTestContext) error {
 
+		if !ctx.SkipStatus {
+			// Duplicate the schema as a status schema
+			statusSchema := sch.DeepCopy()
```
unnecessary deepcopy, no?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go:
```diff
@@ -228,10 +279,21 @@ func (u updateMyCRDV1Beta1Schema) Do(ctx *ratchetingTestContext) error {
+			*v.Subresources = apiextensionsv1.CustomResourceSubresources{
+				Status: &apiextensionsv1.CustomResourceSubresourceStatus{},
+			}
+
```
why set it during schema patching?

JoelSpeed committed:
Simplify status subresource ratcheting testing

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go:
```diff
@@ -174,24 +180,31 @@ func (a applyPatchOperation) Do(ctx *ratchetingTestContext) error {
+	c := ctx.DynamicClient.Resource(a.gvr).Namespace(patch.GetNamespace())
+	if ctx.StatusSubresource {
+		if _, err := c.Get(context.TODO(), patch.GetName(), metav1.GetOptions{}); apierrors.IsNotFound(err) {
+			// ApplyStatus will not automatically create an object, we must make sure it exists before we can
```
nit: any deeper reason why we use Apply in the first place for creations? A `createOperation` might be clearer.

JoelSpeed replied:
It's how I found it to be honest.

For the non subresource path, if you apply an object it auto-creates the object if it doesn't exist, or updates it if it does. So I guess that was cleaner since this operation is used throughout tests not only to create objects, but also update them.

Would you prefer to see a `createOperation` that is used in all tests to make sure the object exists? I guess it would create the object with an empty spec/status and then all apply operations would definitely be updates, not sometimes creates

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go:
```diff
@@ -1387,7 +1452,8 @@ func TestRatchetingFunctionality(t *testing.T) {
 		{
-			Name: "CEL_transition_rules_should_not_ratchet",
+			Name:       "CEL_transition_rules_should_not_ratchet",
+			SkipStatus: true, // Adding a broken status transition rule prevents the object from being updated.
```
I understand this, why it is skipped

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go:
```diff
@@ -164,7 +165,12 @@ func (a applyPatchOperation) Do(ctx *ratchetingTestContext) error {
+		// Copy the map at the top level to avoid modifying the original.
+		for k, v := range obj {
+			patch.Object[k] = v
+		}
```
remind me why we need this change now?

JoelSpeed replied:
The map is coming in from the test case.

We first run the test case against spec, and then status.

When we run it against spec, we (below) write some extra fields to it (apiVersion, kind, metadata), which updates the map within the test case.

When we then run it against status, those fields are carried over from the spec test, and end up as status fields (status.apiVersion, status.kind, status.metadata), which then get rejected as unknown fields by the API

By coping the data out of the map, we prevent this pollution between the spec and status test case

sttts replied:
Ack. Makes sense. There is `runtime.DeepCopyJSON` for that purpose. Would clarify and simplify what we are doing here.

JoelSpeed replied:
I tried using a DeepCopy already, but any deep copy combination I end up with `panic: cannot deep copy int`, so I think we need to leave this as is.

The alternative I guess is to serialise and deserialise the string to create a copy, is that simpler than just copying the map out?

sttts replied:
you have to type cast the ints in the test cases to int64. Technically, these are invalid now.

JoelSpeed commented:
/retest

cici37 commented:
/triage accepted

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go:
```diff
@@ -219,6 +232,17 @@ func (u updateMyCRDV1Beta1Schema) Do(ctx *ratchetingTestContext) error {
+			}
+		}
+
+		// sentinel must be in the root level of the schema.
```
why?

sttts commented on staging/src/k8s.io/apiextensions-apiserver/test/integration/ratcheting_test.go:
```diff
@@ -219,6 +232,17 @@ func (u updateMyCRDV1Beta1Schema) Do(ctx *ratchetingTestContext) error {
 			sch.Properties = map[string]apiextensionsv1.JSONSchemaProps{}
 		}
```
This block belongs to the code starting with `uuidString := string(uuid.NewUUID())`, at least the go doc suggests so

JoelSpeed committed:
Simplify schema sentinel subresource logic

JoelSpeed committed:
Use DeepCopyJSON to copy testcase input

Unstructured types must use int64, all integers in the test case
must be explicitly cast as int64.

sttts commented:
/lgtm
/approve

sttts commented:
/retest

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/status_strategy.go:
```diff
@@ -94,8 +100,17 @@ func (a statusStrategy) ValidateUpdate(ctx context.Context, obj, old runtime.Obj
+	var correlatedObject *common.CorrelatedObject
+	if utilfeature.DefaultFeatureGate.Enabled(apiextensionsfeatures.CRDValidationRatcheting) {
+		correlatedObject = common.NewCorrelatedObject(uNew.Object, uOld.Object, &model.Structural{Structural: a.structuralSchema})
+		options = append(options, validation.WithRatcheting(correlatedObject.Key("status")))
+		celOptions = append(celOptions, cel.WithRatcheting(correlatedObject))
```
it looks like we can encounter CRDs that said they wanted a status subresource but didn't define a status property, and this leads to a panic in that case

can we detect / guard against that in correlatedObject or here somehow?

xref https://github.com/kubernetes/kubernetes/pull/133721 https://github.com/kubernetes/kubernetes/issues/133651
